### ISSUE TITLE
log probability of Truncated Normal gives NAN as gradient

### ISSUE THREAD

**Issue by Nephalen** [tensorflow/probability#518]
Hi,

I'm currently trying to model a latent bivariable vector using truncated normal distribution. My current approach is using two Truncated Normal distributions that are combined into a single Independent distribution. The base model is a neural network. But I'm getting NAN gradient and I think the reason is similar to the situation below.

For the sample code below,

`tpd_mu = tf.constant([2, 3])`
`tpd_sig = tf.constant([1e-4, 0.07])`
`tpd = tfp.distributions.Independent(tfp.distributions.TruncatedNormal(tpd_mu, tpd_sig, [0, 0], [1, 1]), reinterpreted_batch_ndims = 1)`
`tpd_log = tpd.log_prod(tpd.sample(1000))`
`tpd_grad = tf.gradients(tpd_log, [tpd_mu, tpd_sig])`

`tpd_log` will have a real value log pdf output but `tpd_grad ` will return NAN.

Mathematically, it make sense since since the distribution's moments are too far away from truncate boundary. The log probability of a single variable sample goes to infinity. 

The mean and standard deviation are output of neural network. Is there a way to avoid such situation?

**Comment by srvasude** [comment]
I am closing this, since I am not able to reproduce this. I slightly modified this example by changing tpd_mu and the low and high parameters to float32 (in TF those will automatically be integer parameters), and used a gradient tape. The gradients outputted seemed reasonable.

**Comment by kmdalton** [comment]
@srvasude , I am having trouble with the `TruncatedNormal` gradients too. The following simple example gives me this output:

```
Distribution parameters:
loc   : <tf.Variable 'Variable:0' shape=() dtype=float32, numpy=-6.0>
scale : <TransformedVariable: dtype=float32, shape=[], fn="softplus", numpy=1.0>
low   : 0.0  <dtype: 'float32'>
high  : 10000000000.0 <dtype: 'float32'>

Sample:
z = 0.14593791961669922

Log prob:
q.log_prob(z) = 0.9315528869628906

Gradient values:
  dl/dloc = nan
dl/dscale = nan
```

Example source code:
```python
import tensorflow_probability as tfp
import numpy as np
from tensorflow_probability import distributions as tfd
from tensorflow_probability import bijectors as tfb
import tensorflow as tf


dtype = tf.float32

loc = -6.0
scale = 1.
low = 0.
high = 1e10

q = tfd.TruncatedNormal(
    tf.Variable(loc, dtype=dtype), 
    tfp.util.TransformedVariable(scale, tfb.Softplus(), dtype=dtype),
    tf.convert_to_tensor(low, dtype=dtype), 
    tf.convert_to_tensor(high, dtype=dtype), 
)

with tf.GradientTape() as tape:
    z = q.sample()
    l = q.log_prob(z)

grads = tape.gradient(l, q.trainable_variables)


print(f"""
Distribution parameters:
loc   : {q._loc}  
scale : {q._scale} 
low   : {q._low}  {q._low.dtype}
high  : {q._high} {q._high.dtype}

Sample:
z = {z}

Log prob:
q.log_prob(z) = {l}

Gradient values:
  dl/dloc = {grads[0]}
dl/dscale = {grads[1]}
""")
```

Fixing the `high=1e10` and `low=0.`, I find that the gradient tape returns `nan` for sample gradients if the distribution has a `loc/scale` ratio less than about `-5.24`. 
![image](https://user-images.githubusercontent.com/2790777/97769600-b2e26c80-1b02-11eb-8882-0434e8571a9f.png)
![image](https://user-images.githubusercontent.com/2790777/97769606-bbd33e00-1b02-11eb-97c3-f3e051012572.png)
Wherein the empty space is the region where gradients wrt `loc` and `scale` are invariably `nan`. 

Source code for completeness:  
```python
from matplotlib import pyplot as plt
import tensorflow_probability as tfp
import numpy as np
from tqdm import tqdm
from tensorflow_probability import distributions as tfd
from tensorflow_probability import bijectors as tfb
import tensorflow as tf
import pandas as pd


dtype = tf.float32

scale = 1.
low = 0.
high = 1e10

df = pd.DataFrame()

scale, loc = np.meshgrid(np.linspace(0.1, 5., 100), np.linspace(-5.5, 5., 100))
q = tfd.TruncatedNormal(
    tf.Variable(loc, dtype=dtype), 
    tfp.util.TransformedVariable(scale, tfb.Softplus(), dtype=dtype),
    tf.convert_to_tensor(low, dtype=dtype), 
    tf.convert_to_tensor(high, dtype=dtype), 
)

with tf.GradientTape() as tape:
    z = q.sample()
    l = q.log_prob(z)

grads = tape.gradient(l, q.trainable_variables)



plt.figure()
plt.title("Location Parameter Gradients")
plt.contourf(loc, scale, grads[0])
plt.colorbar(label="dlogp/dloc")
plt.xlabel('loc')
plt.ylabel('scale')

slope = (scale/loc)[np.isnan(grads[0])].min()
x = np.linspace(*plt.xlim(), 100)
plt.plot(x, slope * x, '--r', label=f"loc/scale = {1/slope}", scaley=False, scalex=False)
plt.legend()



plt.figure()
plt.title("Scale Parameter Gradients")
plt.contourf(loc, scale, grads[1])
plt.colorbar(label="dlogp/dscale")
plt.xlabel('loc')
plt.ylabel('scale')


slope = (scale/loc)[np.isnan(grads[0])].min()
x = np.linspace(*plt.xlim(), 100)
plt.plot(x, slope * x, '--r', label=f"loc/scale = {1/slope}", scaley=False, scalex=False)
plt.legend()
plt.show()
```

**Comment by hardianlawi** [comment]
Hi, could we reopen this issue, I'm experiencing the same problem.

**Comment by hardianlawi** [comment]
@srvasude Could we reopen this? Below is another example that you could try running locally to reproduce the problem.

```
import tensorflow_probability as tfp
import tensorflow as tf

mu = tf.Variable(0.0)
sigma = tf.constant(1.0)
lower = tf.constant(6.0)
upper = lower + 5
x = lower + 2

with tf.GradientTape() as tape:
    dist = tfp.distributions.TruncatedNormal(mu, sigma, lower, upper)
    log_prob = dist.log_prob(x)
print(tape.gradient(log_prob, [mu]))
```

I would like to know a workaround for this problem. I tried to use `tf.where` as a workaround by setting a threshold to decide  whether to take the gradient of the TruncatedNormal or not, however, there's an issue with the `tf.where` itself. https://github.com/tensorflow/tensorflow/issues/38349

**Comment by Joshuaalbert** [comment]
#1297 Might be related
